core/src/main/java/org/elasticsearch/index/shard/TranslogOpToEngineOpConverter.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.shard;
+
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.mapper.DocumentMapperForType;
+import org.elasticsearch.index.mapper.MapperService;
+import org.elasticsearch.index.translog.Translog;
+
+import static org.elasticsearch.index.mapper.SourceToParse.source;
+
+/**
+ * The TranslogOpToEngineOpConverter encapsulates all the logic needed to transform a translog entry into an
+ * indexing operation including source parsing and field creation from the source.
+ */
+public class TranslogOpToEngineOpConverter {
+    private final MapperService mapperService;
+    private final Logger logger;
+    private final ShardId shardId;
+
+    protected TranslogOpToEngineOpConverter(ShardId shardId, MapperService mapperService, Logger logger) {
+        this.shardId = shardId;
+        this.mapperService = mapperService;
+        this.logger = logger;
+    }
+
+    protected DocumentMapperForType docMapper(String type) {
+        return mapperService.documentMapperWithAutoCreate(type); // protected for testing
+    }
+
+    public Engine.Operation convertToEngineOp(Translog.Operation operation, Engine.Operation.Origin origin) {
```

Comment:
this class seems simple enough to be unittested? Maybe we can add a test based on `IndexShardTestCase`